cdap-app-fabric/src/main/java/io/cdap/cdap/support/task/SupportBundlePipelineInfoTask.java

```diff
@@ -0,0 +1,159 @@
+  private final RemoteProgramLogsFetcher remoteProgramLogsFetcher;
+  private final RemoteMetricsSystemClient remoteMetricsSystemClient;
+  private final List<NamespaceId> namespaces;
+  private final String uuid;
+  private final String application;
+  private final ProgramType programType;
+  private final String programName;
+  private final SupportBundleJob supportBundleJob;
+  private final int maxRunsPerProgram;
+
+  @Inject
+  public SupportBundlePipelineInfoTask(String uuid, List<NamespaceId> namespaces, String application, File basePath,
+                                       RemoteApplicationDetailFetcher remoteApplicationDetailFetcher,
+                                       RemoteProgramRunRecordsFetcher remoteProgramRunRecordsFetcher,
+                                       RemoteProgramLogsFetcher remoteProgramLogsFetcher, ProgramType programType,
+                                       String programName, RemoteMetricsSystemClient remoteMetricsSystemClient,
+                                       SupportBundleJob supportBundleJob, int maxRunsPerProgram) {
+    this.uuid = uuid;
+    this.basePath = basePath;
+    this.namespaces = namespaces;
+    this.application = application;
+    this.remoteApplicationDetailFetcher = remoteApplicationDetailFetcher;
+    this.remoteProgramRunRecordsFetcher = remoteProgramRunRecordsFetcher;
+    this.remoteProgramLogsFetcher = remoteProgramLogsFetcher;
+    this.programType = programType;
+    this.programName = programName;
+    this.remoteMetricsSystemClient = remoteMetricsSystemClient;
+    this.supportBundleJob = supportBundleJob;
+    this.maxRunsPerProgram = maxRunsPerProgram;
+  }
+
+  @Override
+  public void collect() throws IOException, NotFoundException {
+    for (NamespaceId namespaceId : namespaces) {
+      List<ApplicationRecord> appRecords = new ArrayList<>();
+      if (application == null) {
+        appRecords.addAll(remoteApplicationDetailFetcher.list(namespaceId.getNamespace())
+                            .stream()
+                            .map(applicationDetail -> new ApplicationRecord(applicationDetail))
+                            .collect(Collectors.toList()));
+      } else {
+        appRecords.add(new ApplicationRecord(
+          remoteApplicationDetailFetcher.get(new ApplicationId(namespaceId.getNamespace(), application))));
+      }
+
+      for (ApplicationRecord appRecord : appRecords) {
+        String application = appRecord.getName();
+        ApplicationId applicationId = new ApplicationId(namespaceId.getNamespace(), application);
+        //Handle not found for applicationId
+        ApplicationDetail applicationDetail = remoteApplicationDetailFetcher.get(applicationId);
+
+        File appFolderPath = new File(basePath, appRecord.getName());
+        DirUtils.mkdirs(appFolderPath);
+        try (FileWriter file = new FileWriter(new File(appFolderPath, appRecord.getName() + ".json"))) {
+          gson.toJson(applicationDetail, file);
```

Comment:
Close try here to ensure file is closed nicely even if run collection has issues